token/program-2022/src/extension/confidential_transfer/processor.rs

```diff
@@ -800,13 +806,269 @@ fn process_allow_balance_credits(
+    }
+
+    // withheld amount ciphertext must match in the proof data and mint
+    if proof_data.ciphertext_withdraw_withheld_authority
+        != confidential_transfer_mint.withheld_amount
+    {
+        return Err(TokenError::ConfidentialTransferBalanceMismatch.into());
+    }
+
+    // The proof data contains the mint withheld amount encrypted under the destination ElGamal pubkey.
+    // This amount should be added to the destination pending balance.
+    let new_dest_pending_balance = ops::add(
+        &dest_confidential_transfer_account.pending_balance,
+        &proof_data.ciphertext_dest,
+    )
+    .ok_or(ProgramError::InvalidInstructionData)?;
+
+    dest_confidential_transfer_account.withheld_amount = new_dest_pending_balance;
+
+    // fee is now withdrawn, so zero out mint withheld amount
+    confidential_transfer_mint.withheld_amount = EncryptedWithheldAmount::zeroed();
+
+    Ok(())
+}
+
+fn process_withdraw_withheld_tokens_from_accounts(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    num_token_accounts: u8,
+    proof_instruction_offset: i64,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let mint_account_info = next_account_info(account_info_iter)?;
+    let dest_account_info = next_account_info(account_info_iter)?;
+    let instructions_sysvar_info = next_account_info(account_info_iter)?;
+    let authority_info = next_account_info(account_info_iter)?;
+    let authority_info_data_len = authority_info.data_len();
+    let account_infos = account_info_iter.as_slice();
+    let num_signers = account_infos
+        .len()
+        .saturating_sub(num_token_accounts as usize);
+
+    let mut mint_data = mint_account_info.data.borrow_mut();
+    let mut mint = StateWithExtensionsMut::<Mint>::unpack(&mut mint_data)?;
+
+    // mint must be extended for fees
+    let fee_mint = mint.get_extension::<TransferFeeConfig>()?;
+    let withdraw_withheld_authority = Option::<Pubkey>::from(fee_mint.withdraw_withheld_authority)
+        .ok_or(TokenError::NoAuthorityExists)?;
+    Processor::validate_owner(
+        program_id,
+        &withdraw_withheld_authority,
+        authority_info,
+        authority_info_data_len,
+        &account_infos[..num_signers],
+    )?;
+
+    let mut dest_account_data = dest_account_info.data.borrow_mut();
+    let mut dest_account = StateWithExtensionsMut::<Account>::unpack(&mut dest_account_data)?;
+    if dest_account.base.mint != *mint_account_info.key {
+        return Err(TokenError::MintMismatch.into());
+    }
+    if dest_account.base.is_frozen() {
+        return Err(TokenError::AccountFrozen.into());
+    }
+
+    // sum up the withheld amounts in all the accounts
+    let mut aggregate_withheld_amount = EncryptedWithheldAmount::zeroed();
+    for account_info in &account_infos[num_signers..] {
+        // self-harvest, can't double-borrow the underlying data
+        if account_info.key == dest_account_info.key {
+            dest_account
```

Comment:
I don't think checking for the `TransferFeeAmount` extension here is necessary

Reply:
Agreed.